src/rust-crypto/dehydration.ts

```diff
@@ -0,0 +1,262 @@
+                    prefix: UnstablePrefix,
+                },
+            );
+        } catch (error) {
+            const err = error as MatrixError;
+            // We ignore M_NOT_FOUND (there is no dehydrated device, so nothing
+            // us to do) and M_UNRECOGNIZED (the server does not understand the
+            // endpoint).  We pass through any other errors.
+            if (err.errcode === "M_NOT_FOUND" || err.errcode === "M_UNRECOGNIZED") {
+                this.logger.info("dehydration: No dehydrated device");
+                return false;
+            }
+            throw err;
+        }
+
+        this.logger.info("dehydration: device found");
+
+        const rehydratedDevice = await this.olmMachine
+            .dehydratedDevices()
+            .rehydrate(
+                this.key!,
+                new RustSdkCryptoJs.DeviceId(dehydratedDeviceResp.device_id),
+                JSON.stringify(dehydratedDeviceResp.device_data),
+            );
+
+        this.logger.info("dehydration: device rehydrated");
+
+        let nextBatch: string | undefined = undefined;
+        let toDeviceCount = 0;
+        let roomKeyCount = 0;
+        for (;;) {
+            const path = encodeUri("/dehydrated_device/$device_id/events", {
+                $device_id: dehydratedDeviceResp.device_id,
+            });
+            const eventResp: IDehydratedDeviceEventsResp = await this.http.authedRequest<IDehydratedDeviceEventsResp>(
+                Method.Post,
+                path,
+                undefined,
+                nextBatch ? { next_batch: nextBatch } : {},
+                {
+                    prefix: UnstablePrefix,
+                },
+            );
+
+            if (eventResp.events.length == 0) {
+                break;
+            }
+            toDeviceCount += eventResp.events.length;
+            nextBatch = eventResp.next_batch;
+            const roomKeyInfos: RustSdkCryptoJs.RoomKeyInfo[] = await rehydratedDevice.receiveEvents(
+                JSON.stringify(eventResp.events),
+            );
+            roomKeyCount += roomKeyInfos.length;
+
+            // FIXME: is this actually needed?  It looks like the OlmMachine
+            // automatically re-tries decryption
+            await this.rustCrypto.onRoomKeysUpdated(roomKeyInfos);
+        }
+        this.logger.info(`dehydration: received ${roomKeyCount} room keys from ${toDeviceCount} to-device events`);
+
+        return true;
+    }
+
+    /**
+     * Creates and uploads a new dehydrated device
+     *
+     * Creates and stores a new key in secret storage if none is available.
+     */
+    public async createAndUploadDehydratedDevice(): Promise<void> {
+        await this.getKey(true);
+
+        // FIXME: should raise error if server doesn't support dehydration?
+
+        const dehydratedDevice = await this.olmMachine.dehydratedDevices().create();
+        // FIXME: should the device display name be localised? passed as a
+        // parameter?
```

Comment:
I'm not sure this FIXME is actionable. Realistically, should we just accept the fact that we are doing this, or otherwise, log an issue about it somewhere and link from here?

Reply:
removed the FIXME, will create an issue